Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h

```diff
@@ -1024,9 +1044,18 @@ AirIRGeneratorBase<Derived, ExpressionType>::AirIRGeneratorBase(const ModuleInfo
                 // stack that such a stub would use.
                 Checked<uint32_t>(m_maxNumJSCallArguments) * sizeof(Register) + jsCallingConvention().headerSizeInBytes
             ));
-            const int32_t checkSize = m_makesCalls ? (wasmFrameSize + extraFrameSize).value() : wasmFrameSize.value();
+            int32_t checkSize = wasmFrameSize.value();
+            bool wasmFrameSizeNeedsOverflowCheck = wasmFrameSize >= static_cast<int32_t>(minimumParentCheckSize);
+
+            // NOTE: a negative tailCallFrameSize implies that we can reduce the extra frame size.
+            if (m_makesCalls)
+                checkSize = checkedSum<int32_t>(checkSize, extraFrameSize).value();
+            else if (!wasmFrameSizeNeedsOverflowCheck && m_makesTailCalls)
+                checkSize = std::max<Checked<int32_t>>(wasmTailCallFrameSize + extraFrameSize, 0).value();
```

Comment:
Shouldn't we add wasmFrameSize here?

It seems like wasmTailCallFrameSize actually represents the maximum additional stack space needed by any call we make. That is, it is the number of additional locals we need to be prepared to have room for. It does not store the total frame size, and does not incorporate any information about the number of locals we have.

In that case, we should just add the tail call argument size to check size if it is positive.

We can never reduce the frame size that we check against, a negative value here only means that none of our callees will ever have to do their own stack check.